firestore.rules

```diff
@@ -0,0 +1,129 @@
+rules_version = '2';
+
+// Read rules can be divided into get and list rules
+// Write rules can be divided into create, update, and delete rules
+
+service cloud.firestore {
+  match /databases/{database}/documents {
+
+    function isSignedIn() {
+      return request.auth != null;
+    }
+    function isVerified() {
+      return isSignedIn() && request.auth.token.email_verified;
+    }
+
+    match /users/{uid} {
+      function isOwner() {
+        return isVerified() && request.auth.uid == uid;
+      }
+      function validateJoinedAssociations() {
+        // Prevent the user from changing the joinedAssociations field
+        // This will be handled by a Firebase Function
+        return request.resource.data.joinedAssociations == resource.data.joinedAssociations;
+      }
+      function validate() {
+        let fields = ['uid', 'email', 'firstName', 'lastName', 'biography', 'savedEvents', 'followedAssociations', 'joinedAssociations', 'interests', 'socials', 'profilePicture'];
+        let hasCorrectNumberOfFields = request.resource.data.size() == fields.size();
+        let hasCorrectFields = request.resource.data.keys().hasAll(fields);
+
+        return hasCorrectNumberOfFields && hasCorrectFields &&
```

Comment:
This boolean is quite difficult to understand. What is it doing ? maybe add a comment or extract it to a function that has a comprehensible name?

Reply:
This entire expression does data validation of a user object. That means that it checks that the uploaded object contains exactly the specific fields defined above, and each field is of the correct type.
These can be improved at a later point in time with even more data validation, for example checking that `profilePicture` is an email coming from our storage bucket, but for now this seems okay to me.

Also, note that Firestore restricts the amount of functions/variables declared in a ruleset in order to limit their processing time. Thus, I believe we should not extract too many expressions in their own functions/variables in order to allow further scaling. See the [Firestore security rules limits](https://firebase.google.com/docs/firestore/security/rules-structure#security_rule_limits).

Reply:
Oh sorry I didn't see that. Of course then doing helper functions might not be needed if there's a limit